src/main/java/net/trollyloki/advancementbingo/BingoManager.java

```diff
@@ -111,8 +119,85 @@ public void onWin(@NotNull BingoTeam team) {
                 Component.empty(),
                 Title.Times.times(Duration.ofMillis(500), Duration.ofSeconds(3), Duration.ofSeconds(1))
         );
-        for (Player player : Bukkit.getOnlinePlayers())
+        for (Player player : Bukkit.getOnlinePlayers()) {
             player.showTitle(title);
+            Jingles.playGameEnd(plugin, player);
+        }
+    }
+
+    private void onBingoHit(BingoTeam team, NamespacedKey advancement) {
+        if (team.getBoard().isEmpty()) return;
+        @NotNull BingoBoard board = team.getBoard().get();
+
+        int[] advancementLocation = board.findAdvancement(advancement);
+        if (advancementLocation == null) return;
+        int row = advancementLocation[0];
+        int col = advancementLocation[1];
+
+        // Hit for {Team name}
+        // Progress on rows/columns/diagonals
+        // Total bingos scored
+
+        Component hitMessage = hitMessage(team);
+        Component lineMessage = progressMessage(board, row, col);
+        Component completedMessage = Component.text("Completed " + board.getCompletedRows() + "/" + board.getRequiredRows() + " bingos to win")
+                .color(team.getTextColor());
+
+        // Send message next tick so it's displayed after vanilla advancement message
+        Bukkit.getScheduler().runTask(plugin, () -> {
+            for (UUID uuid : players.keySet()) {
```

Comment:
Is there a reason this loop uses the players to teams map rather than just the online players like the previous loop in `onWin`?

Reply:
nope - by habit I only sent the messages to the players that are actually in the game, but since this plugin's only intended to be running on a standalone server it doesn't need to be this way